pkg/api/api.go

```diff
@@ -167,6 +167,7 @@ func (a *API) RegisterAlertmanager(am *alertmanager.MultitenantAlertmanager, tar
 	// Ensure this route is registered before the prefixed AM route
 	a.RegisterRoute("/multitenant_alertmanager/status", am.GetStatusHandler(), false, "GET")
 	a.RegisterRoute("/multitenant_alertmanager/ring", http.HandlerFunc(am.RingHandler), false, "GET", "POST")
+	a.RegisterRoute("/multitenant_alertmanager/delete_tenant_config", http.HandlerFunc(am.DeleteUserConfig), true, "POST")
```

Comment:
Two questions:

- Why `POST` instead of `DELETE`? The semantics of the endpoint remains unchanged so it seems strange that we we have different verbs for things that achieve the same thing.
- I find that using `X-Scope-OrgID` as a way to identify the user to delete seems confusing and violates every other use case (correct me if I'm wrong) within Cortex. We use `X-Scope-OrgID` to identify who is making the request, as opposed to the user to remove. I think `/multitenant_alertmanager/delete_tenant_config/:tenant_id` would suit it better and remove the auth.

Reply:
This comment probably also applies to #3899 but we can move there once we're done with this one.

Reply:
> * Why `POST` instead of `DELETE`? The semantics of the endpoint remains unchanged so it seems strange that we we have different verbs for things that achieve the same thing.

`DELETE` is meant to delete a resource, and while it makes sense for `/api/v1/alert` which also supports `GET` and `POST`, I don't think it does here — `/multitenant_alertmanager/delete_tenant_config` doesn't behave as a resource with other methods available. That is why I chose `POST` here.

> * I find that using `X-Scope-OrgID` as a way to identify the user to delete seems confusing and violates every other use case (correct me if I'm wrong) within Cortex. We use `X-Scope-OrgID` to identify who is making the request, as opposed to the user to remove. I think `/multitenant_alertmanager/delete_tenant_config/:tenant_id` would suit it better as long as we remove the auth (which could serve its own conversation but I'd expect it to be).

This follows design for `/purger/delete_tenant`, as discussed [in the proposal](https://cortexmetrics.io/docs/proposals/tenant-deletion/#post-purgerdelete_tenant). Idea was to keep it consistent with rest of Cortex, and avoid introducing new mechanism to identify tenant. It also works nicely with "fake" user when authentication is disabled.

Reply:
> This follows design for /purger/delete_tenant, as discussed in the proposal. Idea was to keep it consistent with rest of Cortex, and avoid introducing new mechanism to identify tenant. It also works nicely with "fake" user when authentication is disabled.

I know it's a bit late to take comment on an already established proposal so please take it with a grain of salt.

In my eyes, this has the potential to be disastrous. From an operator point of view, we end to issue calls against the Cortex API with an `X-Scope-OrgID` set when trying to impersonate a tenant. Having the same semantics to actually delete a resource seems unorthodox and counter-intuitive.

I don't want to be a nuisance or impose a block on this work, but I thought it was worth mentioning.

Reply:
> In my eyes, this has the potential to be disastrous. From an operator point of view, we end to issue calls against the Cortex API with an X-Scope-OrgID set when trying to impersonate a tenant. Having the same semantics to actually delete a resource seems unorthodox and counter-intuitive.

I'm not sure I fully follow what you're saying here.

If we used `/multitenant_alertmanager/delete_tenant_config/:tenant_id`, perhaps it would be nicer, but it's still deleting config.

Reply:
What I'm saying is that it's easy to miss that `X-Scope-OrgID` is used to identify who you are deleting vs who is executing the call, but I guess it's no different from issuing the call against the `DELETE /api/v1/alerts` directly.

Reply:
> What I'm saying is that it's easy to miss that X-Scope-OrgID is used to identify who you are deleting vs who is executing the call

I see what you mean here but in Cortex there's no "who you are deleting vs who is executing the call" difference. Cortex expects any request to be previously authenticated/authorized and the request `X-Scope-OrgID` should contain the ID of the tenant upon which Cortex should operate. In this regards, what Peter did looks consistent with the rest of Cortex to me.

Reply:
All good, just thought it was worth mentioning it.